### PR TITLE
Improve our testing of DefaultAnalyzerAssemblyLoader

### PR DESCRIPTION
**Pull request by jaredpar**
As I was working on the fix for #66104 I noticed that our tests for `DefaultAnalyzerAssemblyLoader` weren't actually behaving the way they should. The reasoning is that all of the tests in the class ran inside the same `AppDomain / AssemblyLoadContext`. These tests are all about how we load assemblies and manage them. Most assumed they were starting from a clean state but in reality were being polluted by the tests that ran before them. Specifically ...

For .NET Framework there is just one `AppDomain`. Every assembly was being loaded into that `AppDomain`. It meant that implicit loads during tests were getting results from the previous test thus impacting the outcome. Even worse is `DefaultAnalyzerAssemblyLoader` hooks `AppDomain.AssemblyResolve` so new tests were potentially getting interferred with by previous `DefaultAnalyzerAssemblyLoader` instances resolve functions. 

For .NET Core the same compiler `AssemblyLoadContext` was used between tests. That had similar pollution problems. It also meant that all of our tests were validating our behavior when the `AssemblyLoadContext` for the compiler was the same as the default `AssemblyLoadContext`. That is definitely true for command line invocations of the compiler but is not true when the compiler is hosted in Visual Studio. The behavior there is different and it has resulted in bugs in the past. 

This change fixes the issues here by forcing the tests to execute on a new `AppDomain / AssemblyLoadContext`. The change has two commits to help with reviewing: 

- Commit 1 is simply the structural change to use separate `AppDomain / AssemblyLoadContext` per test. It's a bit of setup code + lots of mechanical changes to let the code execute in a different context. 
- Commit 2 is fixing up the test failures that resulted from the increased isolation in the tests.

### CODE REVIEW

**Comment by jaredpar** [comment]
@dotnet/roslyn-compiler, @RikkiGibson FYI

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 1]
```diff
@@ -28,642 +37,846 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
         }
 
         [Fact, WorkItem(32226, "https://github.com/dotnet/roslyn/issues/32226")]
         public void LoadWithDependency()
         {
-            var analyzerDependencyFile = _testFixture.AnalyzerDependency;
-            var analyzerMainFile = _testFixture.AnalyzerWithDependency;
-            var loader = new DefaultAnalyzerAssemblyLoader();
-            loader.AddDependencyLocation(analyzerDependencyFile.Path);
+            Run(static (DefaultAnalyzerAssemblyLoader loader, AssemblyLoadTestFixture testFixture) =>
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:111]: Did you verify that when the delegate throws an exception that it propagates correctly back to here?

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 2]
```diff
@@ -28,642 +37,846 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
+                    ("Gamma", "0.0.0.0", gammaFile.Path)
+                );
+
 #endif
+            });
         }
 
+        /// <summary>
+        /// Similar to <see cref="AssemblyLoading_DependencyInDifferentDirectory"/> except want to validate
+        /// a dependency in the same directory is preferred over one in a different directory.
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:276]: You're certain we didn't already have this test? this sounds very very familiar. Also, it might be good to choose one of the Delta1 variants which has different output at runtime. Otherwise, I think it's difficult to be sure if the dependency in the same directory really was preferred when running on framework.

Although I'd imagine there's also a moment-in-time problem when on framework, if another analyzer loaded something with a matching name, there's no opportunity for this analyzer to "override" it.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:111]: Indeed yes I verified the exceptions propagate, trigger errors and have meaningful error messages in the test explorer. That's how I went about fixing the broken tests that I found.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 2 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:276]: > Also, it might be good to choose one of the Delta1 variants which has different output at runtime

In this case I wanted to use the same DLL in two places because I wanted version to be identical. Didn't want version greater than coming into play. 

> Otherwise, I think it's difficult to be sure if the dependency in the same directory really was preferred when running on framework.

Actually, now that we have isolated `AppDomain` per run we can basically do the same verification we do for .NET Core. We can check the loaded assemblies. Let me do that.

File: src/Compilers/Test/Core/Platform/CoreClr/AssemblyLoadContextUtils.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        public static SimpleAssemblyLoadContext Create(string name, string? probingPath = null) =>
+            new SimpleAssemblyLoadContext(name, probingPath);
+    }
+
+    public sealed class SimpleAssemblyLoadContext : AssemblyLoadContext
+    {
+        private readonly string _probingPath;
+
+        public SimpleAssemblyLoadContext(string name, string? probingPath = null)
```
Comment (jcouv) [inline src/Compilers/Test/Core/Platform/CoreClr/AssemblyLoadContextUtils.cs:30]: It looks like `probingPath` is always `null` (couldn't find a caller that passes an actual value). Or did I miss something?

File: src/Compilers/Test/Core/Platform/CoreClr/AssemblyLoadContextUtils.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if NETCOREAPP
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
```
Comment (jcouv) [inline src/Compilers/Test/Core/Platform/CoreClr/AssemblyLoadContextUtils.cs:10]: Looks like there are some unnecessary usings

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 5]
```diff
@@ -28,127 +37,195 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
         private AssemblyLoadTestFixtureCollection() { }
     }
 
-    [Collection(AssemblyLoadTestFixtureCollection.Name)]
     public sealed class DefaultAnalyzerAssemblyLoaderTests : TestBase
     {
+#if NETCOREAPP
+        private sealed class AssemblyLoadContextInvokeUtil
```
Comment (jcouv) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:43]: Could we remove this type and use `AppDomainInvokeUtil` for both platforms instead?

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 6]
```diff
@@ -28,642 +37,846 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
+                Assembly analyzerAssembly = loader.LoadFromPath(testFixture.AnalyzerReferencesSystemCollectionsImmutable2.Path);
+                var analyzer = analyzerAssembly.CreateInstance("Analyzer")!;
+                analyzer.GetType().GetMethod("Method")!.Invoke(analyzer, new object[] { sb });
+                Assert.Equal(ExecutionConditionUtil.IsCoreClr ? "1" : "42", sb.ToString());
+            });
         }
 
-        [ConditionalFact(typeof(WindowsOnly), typeof(CoreClrOnly))]
-        public void AssemblyLoading_NativeDependency()
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/66104")]
```
Comment (jcouv) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:810]: Why skip?

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 7]
```diff
@@ -28,642 +37,846 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
+                Assert.Same(copiedAssembly, assembly);
+            });
+        }
 
-            Assembly analyzerAssembly = loader.LoadFromPath(_testFixture.AnalyzerWithNativeDependency.Path);
-            var analyzer = analyzerAssembly.CreateInstance("Class1")!;
-            var result = analyzer.GetType().GetMethod("GetFileAttributes")!.Invoke(analyzer, new[] { _testFixture.AnalyzerWithNativeDependency.Path });
-            Assert.Equal(0, Marshal.GetLastWin32Error());
-            Assert.Equal(FileAttributes.Archive, (FileAttributes)result!);
+        [ConditionalFact(typeof(WindowsOnly))]
```
Comment (jcouv) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:830]: I didn't follow why this test is not longer applicable to Core. Can you clarify?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 3)

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:111]: Ah I see the confusion. I after writing the code to do this, and using it to debug test failures, I accidentally deleted it when cleaning up before pushing. Will add back shortly.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 6 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:810]: Issue #66104 is what caused me to start digging into this and finding the issue in our test harness. Originally I had the fix for #66104 and the test fix in the same PR. After some thought decided to break them into diff PRs to make review easier. This is just disabled until the next PR.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 7 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:830]: Previously the test only ran on Windows CoreClr, now it runs on Windows CoreClr or Framework. It's an expansion of the test.

File: src/Compilers/Test/Core/Platform/CoreClr/AssemblyLoadContextUtils.cs (hunk 3 above)
Comment (jaredpar) [reply src/Compilers/Test/Core/Platform/CoreClr/AssemblyLoadContextUtils.cs:30]: It is `null` in our current use. This is a basic utility class I copied over though which supports alternate probing paths. It's a common helper for `AssemblyLaodContext` usage. Didn't want to delete it only to have to add it back later when I needed it.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:326]: > In this case I wanted to use the same DLL in two places because I wanted version to be identical.

IIRC, the test fixture contains multiple "versions" of Delta with exactly the same assembly version but different behavior. That's what I had in mind. But, if we can verify the paths of the loaded assemblies on framework, it's not essential to do this.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 5)

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 2 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:326]: Yes indeed Delta2B and Delta2 have the same version but different content on a different path. The newest version of the change though allows us to verify paths on both .NET Framework and .NET Core. The .NET Framework code path is less ideal since we don't have the isolation of `AssemblyLoadContext` but I found a way to tease out these values.

**Comment by jaredpar** [comment]
@RikkiGibson believe I've addressed all your feedback here.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 2 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:339]: Actually I found a method to detect if an `Assembly` is in the _LoadFrom_ context vs. _Load_ in .NET Framework so we can be very direct with our tests here now.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 8]
```diff
@@ -28,642 +37,878 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
+            public void Exec(string typeName, string methodName)
+            {
+                try
+                {
+                    InvokeTestCode(typeName, methodName);
+                }
+                catch (TargetInvocationException ex) when (ex.InnerException is XunitException)
+                {
+                    var inner = ex.InnerException;
+                    throw new Exception(inner.Message + inner.StackTrace);
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:59]: Is there a tradeoff between throwing this new exception and doing `throw inner` here?

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 9]
```diff
@@ -28,642 +37,878 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
+        public DefaultAnalyzerAssemblyLoaderTests(ITestOutputHelper output)
         {
             _output = output;
-            _testFixture = testFixture;
+        }
+
+        private void Run(Action<DefaultAnalyzerAssemblyLoader, AssemblyLoadTestFixture> action, [CallerMemberName] string? memberName = null)
+        {
+#if NETCOREAPP
+            var alc = AssemblyLoadContextUtils.Create($"Test {memberName}");
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:76]: Could you explain what test gaps we are filling in by doing this indirection on the netcore side? I thought we already didn't have to worry about analyzer assemblies being loaded in any ALC which persists across tests.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 10]
```diff
@@ -28,642 +37,878 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
         }
 
         private static void VerifyAssemblies(IEnumerable<Assembly> assemblies, params (string simpleName, string version, string path)[] expected)
         {
             Assert.Equal(expected, Roslyn.Utilities.EnumerableExtensions.Order(assemblies.Select(assembly => (assembly.GetName().Name!, assembly.GetName().Version!.ToString(), assembly.Location))));
         }
 
-        [ConditionalFact(typeof(CoreClrOnly))]
+        /// <summary>
+        /// Verify the set of asesmblies loaded as analyzer dependencies are the specified assembly paths
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:248]:
```suggestion
        /// Verify the set of assemblies loaded as analyzer dependencies are the specified assembly paths
```

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 11]
```diff
@@ -28,642 +37,878 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
+
+#if NETCOREAPP
+            // This verify only works where there is a single load context.
+            var alcs = DefaultAnalyzerAssemblyLoader.TestAccessor.GetOrderedLoadContexts(loader);
+            Assert.Equal(1, alcs.Length);
+
+            loadedAssemblies = alcs[0].Assemblies;
+#else
+
+            // Unfortunately in .NET Framework we cannot determine which Assemblies are in the LoadFrom context
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:262]: It looks like this comment should be revised, since we have a trick for determining if the assembly is in the LoadFrom context.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 12]
```diff
@@ -28,642 +37,878 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
+                    // we will end up bringing these assemblies into the Load context by handling them
+                    // there.
+                    undidHook = loader.EnsureResolvedUnhooked();
+                    var name = assembly.FullName;
+                    var other = AppDomain.CurrentDomain.Load(name);
+                    return other is null || other != assembly;
+                }
+                catch
+                {
+                    return true;
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:284]: Can you help me understand this bit? What in the above try block is going to throw and why does that make us think the assembly is in the LoadFrom context?

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 13]
```diff
@@ -28,642 +37,878 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
                     actual);
-            }
-            else
-            {
-                Assert.Equal(
-@"Delta: Epsilon: Test E
+                }
+                else
+                {
+                    // The Epsilon.dll has Delta.dll (v2) next to it in the directory. The .NET Framework
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:497]: Could you explain the behavior change here a bit more? I am afraid that maybe before this PR the behavior depended on the order of execution of tests 😓

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 14]
```diff
@@ -28,642 +37,878 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
-            loader.AddDependencyLocation(_testFixture.AnalyzerWithNativeDependency.Path);
+            Run(static (DefaultAnalyzerAssemblyLoader loader, AssemblyLoadTestFixture testFixture) =>
+            {
+                var assembly = typeof(ImmutableArray<int>).Assembly;
+
+                // Copy the dependency to a new location to simulate it being deployed by an 
+                // analyzer / generator
+                using var tempRoot = new TempRoot();
+                var destFile = tempRoot.CreateDirectory().CreateOrOpenFile($"{assembly.GetName().Name}.dll");
+                destFile.CopyContentFrom(assembly.Location);
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:853]: We have a test for a fake ImmutableArray dependency already, does this test exercise something new about that scenario?

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs [hunk 15]
```diff
@@ -28,642 +37,878 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
...
+        {
+            Run(static (DefaultAnalyzerAssemblyLoader loader, AssemblyLoadTestFixture testFixture) =>
+            {
+                const int INVALID_FILE_ATTRIBUTES = -1;
+                loader.AddDependencyLocation(testFixture.AnalyzerWithNativeDependency.Path);
+
+                Assembly analyzerAssembly = loader.LoadFromPath(testFixture.AnalyzerWithNativeDependency.Path);
+                var analyzer = analyzerAssembly.CreateInstance("Class1")!;
+                var result = analyzer.GetType().GetMethod("GetFileAttributes")!.Invoke(analyzer, new[] { testFixture.AnalyzerWithNativeDependency.Path });
+                Assert.NotEqual(INVALID_FILE_ATTRIBUTES, result);
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:873]: Can you explain what changed here?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs [hunk 16]
```diff
@@ -33,15 +33,11 @@ internal class DefaultAnalyzerAssemblyLoader : AnalyzerAssemblyLoader
...
-        private int _hookedAssemblyResolve;
+        private bool _hookedAssemblyResolve;
 
         protected override Assembly LoadFromPathUncheckedImpl(string fullPath)
         {
-            if (Interlocked.CompareExchange(ref _hookedAssemblyResolve, 0, 1) == 0)
-            {
-                AppDomain.CurrentDomain.AssemblyResolve += CurrentDomain_AssemblyResolve;
-            }
+            EnsureResolvedHooked();
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs:40]: Does this result in any behavior change?

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 8 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:59]: Yes. `XunitException` is not serializable so it cannot be thrown across `AppDomain` boundaries. The best I could find is to wrap the content here and rethrow. It's usable enough that I was able to debug all the issues that went through this code path.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 9 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:76]: Tried to explain this a bit in the PR description. 

The gap is not as big for `AssemblyLoadContext` because, as you mention, the isolation is already there. There is no cross test pollution to worry about, not significantly at least. 

What this does buy us though is verifying how we handle the default ALC vs. the compiler ALC. Before this change in our tests the default ALC was the compiler ALC while now they are different instances. When they are different is the more complex case and how VS executes. This gives us better confidence we're falling back to the compiler ALC during loads vs. accidentally falling back to default.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 12 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:284]: A call to `AppDomain.Load(string)` will only succeed if one of the following is true: 

1. An `Assembly` with that name is already present in the _Load_ context
2. An `Assembly` with that name was found in the probing path of the `AppDomain`. Generally this is the directory where the application was run from or assemblies in the GAC
4. The `AppDomain.ResolveAssembly` event was able to find an assembly with the specified name. 

When none of these are true an exception is thrown and that demonstrates it's in the _LoadFrom_ context not _Load_. 

Next I expect you'll want to know why I also test for `null`. Mostly for completeness because I can't prove to myself that `null` won't be returned in edge cases :smile:

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 13 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:497]: When you load a DLL via `LoadFrom` the path of that DLL becomes a part of the probing path for that DLL (but not others). In this case the director layout is as follows:

- Version2
    - Epsilon.dll 
    - Delta.dll (version 2) 

The compile loads Epsilon.dll into the _LoadFrom_ context via the `loader.LoadFromPath(testFixture.Epsilon.Path)` call above. That puts the Version2 directory on the private probing path for Epsilon.dll. When the time comes to resolve Delta.dll the runtime notices it's not available in the _LoadFrom_ or _Load_ context so it attempts to load it. Because Delta.dll is on the private probing path this load happens implicitly and the DLL is resolved.

There is no call to `AppDomain.ResolveAssembly` here so no chance for us to intercept this load. It is completely outside our control. Unlike in .NET Core where every load is controlled by an `AssemblyLoadContext` and we can force it to only pick registered dependencies. 

I do suspect this test was working due to prior test pollution before. Prior tests would've put a different version of Delta.dll into _LoadFrom_ and that would've satisfied the implicit load.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 14 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:853]: This is a case where the dependency is identical so it will check every single box when we look for compatibility.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs (hunk 15 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:873]: This change cost me about one hour of my life to track down. The real issue is that `GetLastWin32Error` is only interesting to call when a native interop call fails. When that call succeeds it has unknown data inside of it. That is because a native call only sets the last error on failure, on success it does nothing. 

In the change where I fix the bug I added a `File.Exists` check on this code path. That check fails and sets the last error to 2. No other native call fails on this code path so when we get to this call the `Assert.Equal` failed because of that prior `File.Exists` failure. 

The actual failure code for `GetFileAttributesW` is `INVALID_FILE_ATTRIBUTES` so I moved the check to be that. Really I considered just deleting this entirely as the `Assert.Equal` check below is sufficient to verify this case.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs (hunk 16 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs:40]: The ability to toggle the hook is necessary to determine if an assembly is in the _LoadFrom_ context. If we leave the hook enabled then the call to `Load(string name)` comes through the resolver, we find it in the directories, and then return the loaded assembly. Hence have to be able to turn the hook on / off during testing to probe _LoadFrom_ correctly.

**Review by RikkiGibson — APPROVED** [review]
